### PR TITLE
feat: Implement support for cross-module filtering at the DAL

### PR DESCRIPTION
**Pull request by sradevski**
I am still doing some final reviews and testing, but it should mostly work. Need to test few more edge-cases

### CODE REVIEW

File: packages/core/types/src/dal/cross-module-query.ts [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+export type CrossModuleJoinSpec = {
+  /**
+   * Alias used to reference the external table in order keys (e.g. `customer.email`).
+   * Must be unique across the join specs of a single query.
+   */
+  alias: string
+  /**
+   * When set, this join is correlated to the target row of the named parent
+   * join instead of the root entity. Used for multi-hop filters such as
+   * filtering customers by a property of their linked pricing tier's
```
Comment (peterlgh7) [inline packages/core/types/src/dal/cross-module-query.ts]: there may be more representative examples

File: packages/core/utils/src/dal/mikro-orm/cross-module-query/augment-find-options.ts [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+    defaultSchema,
+    withDeleted,
+  }
+
+  const options = {
+    ...(findOptions.options ?? {}),
+  }
+
+  if (options.__internal) {
+    const internal = { ...options.__internal }
```
Comment (peterlgh7) [inline packages/core/utils/src/dal/mikro-orm/cross-module-query/augment-find-options.ts]: super nit: this could be `const {crossModuleJoins, ...internal} = options.__internal`

File: packages/core/utils/src/dal/mikro-orm/cross-module-query/filter-sql.ts [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+    if (Array.isArray(value)) {
+      const placeholders = value.map(() => "?").join(", ")
+      clauses.push(`${column} in (${placeholders})`)
+      bindings.push(...value)
+      continue
+    }
+
+    if (isObject(value)) {
+      const operatorKeys = Object.keys(value).filter((operator) =>
+        operator.startsWith("$")
```
Comment (peterlgh7) [inline packages/core/utils/src/dal/mikro-orm/cross-module-query/filter-sql.ts:86]: is it safe to ignore the non operator keys here?

File: packages/core/utils/src/dal/mikro-orm/cross-module-query/filter-sql.ts [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+      clauses.push(`${column} is null`)
+      continue
+    }
+
+    if (value === undefined) {
+      continue
+    }
+
+    if (Array.isArray(value)) {
+      const placeholders = value.map(() => "?").join(", ")
```
Comment (peterlgh7) [inline packages/core/utils/src/dal/mikro-orm/cross-module-query/filter-sql.ts]: I think it's safer to pass an array as parameter, cause this may hit the max parameters limit.

File: packages/core/utils/src/dal/mikro-orm/cross-module-query/filter-sql.ts [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+      bindings.push(value as Knex.RawBinding)
+      return `${column} ilike ?`
+    case "$is":
+      if (value === null) {
+        return `${column} is null`
+      }
+      bindings.push(value as Knex.RawBinding)
+      return `${column} is ?`
+    default:
+      if (OPERATOR_MAP[operator]) {
```
Comment (peterlgh7) [inline packages/core/utils/src/dal/mikro-orm/cross-module-query/filter-sql.ts]: what is this default branch covering? Isn't it the same cases as above?

File: packages/core/utils/src/dal/mikro-orm/cross-module-query/order-sql.ts [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+  primaryKey: string,
+  rootAlias: string,
+  defaultSchema: string,
+  withDeleted: boolean
+): FindConfigOrder {
+  const transformed: FindConfigOrder = {}
+
+  for (const [key, direction] of Object.entries(orderBy)) {
+    if (direction === "ASC" || direction === "DESC") {
+      if (!key.includes(".")) {
```
Comment (peterlgh7) [inline packages/core/utils/src/dal/mikro-orm/cross-module-query/order-sql.ts]: Can't we get rid of this special case by doing first `const [alias, ...rest] = key.split(".")` (line 64). I found it confusing to understand the current three codepaths. Looks like it could be 2.

File: packages/core/utils/src/dal/mikro-orm/cross-module-query/order-sql.ts [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+    buildLinkToTargetJoinSql(
+      linkAlias,
+      targetAlias,
+      joinSpec.link.targetKey,
+      targetPrimaryKey
+    ),
+    buildJoinSoftDeleteSql(linkAlias, targetAlias, withDeleted),
+  ].filter(Boolean)
+
+  // For to-many links more than one target row can match. Order by the
```
Comment (peterlgh7) [inline packages/core/utils/src/dal/mikro-orm/cross-module-query/order-sql.ts]: how is this useful for to-many links if the ordered by value ends up being just one out of the joined ones? I can't imagine a situation where I'd order by a to-many field. E.g. ordering orders by payment_collection.amount doesn't make sense to me. I could only imagine more complicated orders like by sum(payment_collection.amount), but that's a more complex use case

File: packages/core/types/src/dal/cross-module-query.ts [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+ *     filters: { email: "user@example.com" },
+ *   },
+ * }
+ * ```
+ *
+ * @internal
+ */
+export type CrossModuleJoinSpec = {
+  /**
+   * Alias used to reference the external table in order keys (e.g. `customer.email`).
```
Comment (peterlgh7) [inline packages/core/types/src/dal/cross-module-query.ts]: why does the query module need to pass the alias? can't the DAL generate it?

File: packages/core/types/src/dal/cross-module-query.ts [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+   */
+  alias: string
+  /**
+   * When set, this join is correlated to the target row of the named parent
+   * join instead of the root entity. Used for multi-hop filters such as
+   * filtering customers by a property of their linked pricing tier's
+   * functionality.
+   */
+  parent?: string
+  link: CrossModuleJoinLink
```
Comment (peterlgh7) [inline packages/core/types/src/dal/cross-module-query.ts:106]: How do you imagine read-only links being handled? Let's imagine table A and table B, with table B having a column `a_id`. We want to query A. I see two options:

1. We keep this interface, and we pass B as `link.table`, B.id as `link.target_key` and B.a_id ad `link.source_key`. This will create an odd_looking query in the DAL, something like `FROM B b1 join B b2 on id = id`. The query planner will probably simplify that and eliminate the join, but it's kinda ugly.
2. We update this interface to something like `link: CrossModuleJoinLink | CrossModuleJoinReadonlyLink`, and we handle readonly links specially in the DAL.

File: packages/core/utils/src/dal/mikro-orm/__tests__/cross-module-query.spec.ts [hunk 10]
```diff
@@ -0,0 +1,848 @@
...
+        whereRaw: jest.fn(function (...args) {
+          calls.push(["whereRaw", ...args])
+          return this
+        }),
+      }
+
+      return qb
+    }
+
+    it("should apply simple equality filters", () => {
```
Comment (peterlgh7) [inline packages/core/utils/src/dal/mikro-orm/__tests__/cross-module-query.spec.ts:37]: nit: I'd prefer tests where we create a full actual query and assert it fully. The queries could be simple enough to make this not too verbose.
But maybe it's not suitable here, just a suggestion.

File: packages/core/utils/src/dal/mikro-orm/integration-tests/__tests__/cross-module-query.spec.ts [hunk 11]
```diff
@@ -0,0 +1,1170 @@
...
+  pgGodCredentials,
+  SqlCapture,
+} from "../__fixtures__/database"
+
+jest.setTimeout(300000)
+
+const dbName = "mikroorm-cross-module-customer-pricing-tier"
+
+@Entity()
+class CustomerEntity {
```
Comment (peterlgh7) [inline packages/core/utils/src/dal/mikro-orm/integration-tests/__tests__/cross-module-query.spec.ts]: Could we not define models and create tables through medusa model definitions? Maybe that's outside the scope of the `dal/mikro-orm` package though.

File: packages/core/types/src/dal/cross-module-query.ts (hunk 1 above)
Comment (sradevski) [reply packages/core/types/src/dal/cross-module-query.ts]: Yeah it started as product-> price, then made it like a pricing tier and it got unnecessarily complex. I'll rework it so it matches some of our modules' models

File: packages/core/utils/src/dal/mikro-orm/cross-module-query/augment-find-options.ts (hunk 2 above)
Comment (sradevski) [reply packages/core/utils/src/dal/mikro-orm/cross-module-query/augment-find-options.ts]: There was some more cleanup, the whole if was unnecessary

File: packages/core/utils/src/dal/mikro-orm/cross-module-query/filter-sql.ts (hunk 3 above)
Comment (sradevski) [reply packages/core/utils/src/dal/mikro-orm/cross-module-query/filter-sql.ts:86]: Yeah from what I saw, all operators have a key, otherwise they refer to fields

File: packages/core/utils/src/dal/mikro-orm/cross-module-query/filter-sql.ts (hunk 5 above)
Comment (sradevski) [reply packages/core/utils/src/dal/mikro-orm/cross-module-query/filter-sql.ts]: Agh leftover, removing

File: packages/core/utils/src/dal/mikro-orm/cross-module-query/filter-sql.ts (hunk 4 above)
Comment (sradevski) [reply packages/core/utils/src/dal/mikro-orm/cross-module-query/filter-sql.ts]: good point, done.

File: packages/core/utils/src/dal/mikro-orm/cross-module-query/order-sql.ts (hunk 6 above)
Comment (sradevski) [reply packages/core/utils/src/dal/mikro-orm/cross-module-query/order-sql.ts]: Yeah good point, done

File: packages/core/types/src/dal/cross-module-query.ts (hunk 8 above)
Comment (sradevski) [reply packages/core/types/src/dal/cross-module-query.ts]: I was a bit undecided here as well, whether the graph of the DAL should do this. Let's do it in the DAL if you feel it's better, I went back and forth on it :D

File: packages/core/types/src/dal/cross-module-query.ts (hunk 9 above)
Comment (sradevski) [reply packages/core/types/src/dal/cross-module-query.ts:106]: Yeah I intentionally didn't tackle read-only links for now (should have added a comment) as they are slightly simpler and wouldn't take much changes to the code. I'd prefer to do it in a separate PR.

File: packages/core/utils/src/dal/mikro-orm/integration-tests/__tests__/cross-module-query.spec.ts (hunk 11 above)
Comment (sradevski) [reply packages/core/utils/src/dal/mikro-orm/integration-tests/__tests__/cross-module-query.spec.ts]: Yeah it's outside the scope, and I didn't want to mix the DML here. I just followed the same logic as other mikroorm tests here

File: packages/core/utils/src/dal/mikro-orm/cross-module-query/order-sql.ts (hunk 7 above)
Comment (sradevski) [reply packages/core/utils/src/dal/mikro-orm/cross-module-query/order-sql.ts]: IIRC when you create a link it creates a pivot table even for entries that are practically 1-to-1, and we did an application-level check.

I agree if it is one-to-many (and the comment was written by cursor, which is a bit wrong) it is not useful, but this just handles the general case. I'll add a comment as it is probably an area for improvement, but we can do it separately IMO

File: packages/core/utils/src/dal/mikro-orm/__tests__/cross-module-query.spec.ts (hunk 10 above)
Comment (sradevski) [reply packages/core/utils/src/dal/mikro-orm/__tests__/cross-module-query.spec.ts:37]: Well I do a more end-to-end assertion in the integration tests, these were meant to be simpler and more isolated.

File: packages/core/utils/src/dal/mikro-orm/cross-module-query/augment-find-options.ts (hunk 2 above)
Comment (peterlgh7) [reply packages/core/utils/src/dal/mikro-orm/cross-module-query/augment-find-options.ts]: right, I thought that too but forgot to comment

File: packages/core/utils/src/dal/mikro-orm/cross-module-query/order-sql.ts (hunk 7 above)
Comment (peterlgh7) [reply packages/core/utils/src/dal/mikro-orm/cross-module-query/order-sql.ts]: Sure, leaving a comment is fine. I just feel this behavior is a bit random and no-one would be counting on it.

File: packages/core/types/src/dal/cross-module-query.ts (hunk 8 above)
Comment (peterlgh7) [reply packages/core/types/src/dal/cross-module-query.ts]: I kinda feel it's better in the DAL, but well, this is all work in progress, there may be reasons to pass it down too.
